pkg/utils/authcheck/authtype.go

```diff
@@ -0,0 +1,139 @@
+/*
+Copyright 2020 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package authcheck
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+
+	"github.com/google/knative-gcp/pkg/reconciler/identity/resources"
+)
+
+type AuthTypes string
+
+type AuthTypeArgs struct {
+	Namespace          string
+	ServiceAccountName string
+	Secret             *corev1.SecretKeySelector
+}
+
+const (
+	AuthenticationCheckUnknownReason = "AuthenticationCheckPending"
+	ControlPlaneNamespace            = "cloud-run-events"
+	BrokerServiceAccountName         = "broker"
+	// Secret option is referring to authentication configuration for secret.
+	// https://cloud.google.com/kubernetes-engine/docs/tutorials/authenticating-to-cloud-platform#importing_credentials_as_a_secret
+	Secret AuthTypes = "secret"
+	// WorkloadIdentityGSA option is referring to authentication configuration for Workload Identity using GSA
+	// https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity
+	WorkloadIdentityGSA AuthTypes = "workload-identity-gsa"
+	WorkloadIdentity    AuthTypes = "workload-identity"
+)
+
+var BrokerSecret = &corev1.SecretKeySelector{
+	LocalObjectReference: corev1.LocalObjectReference{
+		Name: "google-broker-key",
+	},
+	Key: "key.json",
+}
+
+// GetAuthTypeForBrokerCell will get authType for BrokerCell.
+func GetAuthTypeForBrokerCell(ctx context.Context, serviceAccountLister corev1listers.ServiceAccountLister,
+	secretLister corev1listers.SecretLister, args AuthTypeArgs) (AuthTypes, error) {
+	// For AuthTypeArgs from BrokerCell, ServiceAccountName and Secret will be both presented.
+	// We need to revisit this function after https://github.com/google/knative-gcp/issues/1888 lands,
+	// which will add IdentitySpec to BrokerCell.
+	// For AuthTypeArgs from BrokerCell.
+	authTypeForWorkloadIdentity, workloadIdentityErr := getAuthTypeForWorkloadIdentity(ctx, serviceAccountLister, args)
+	authTypeForSecret, secretErr := getAuthTypeForSecret(ctx, secretLister, args)
```

Comment:
The fact that we aren't immediately checking the returned `error` does not match what I normally see in go. It is probably worth commenting on why it is set up this way. 

Alternatively, something like:

```suggestion
	authTypeForWorkloadIdentity, workloadIdentityErr := getAuthTypeForWorkloadIdentity(ctx, serviceAccountLister, args)
	if workloadIdentityErr != nil {
		return authTypeForWorkloadIdentity, nil
	}
	authTypeForSecret, secretErr := getAuthTypeForSecret(ctx, secretLister, args)
	if secretErr != nil {
		return authTypeForSecret, nil
	}
	workloadIdentityError := fmt.Errorf("when checking Kubernetes Service Account %s, got error: %w", args.ServiceAccountName, workloadIdentityErr)
	secretError := fmt.Errorf("when checking Kubernetes Secret %s, got error: %w", args.Secret.Name, secretErr)
	return "", fmt.Errorf("authentication is not configured, %s, %s", workloadIdentityError.Error(), secretError.Error())
```